File: providers/spotify/getSpotifyUser.go
package spotify

import (
	"encoding/json"
)

type SpotifyImage struct {
	Height int    `json:"height"`
	Width  int    `json:"width"`
	Url    string `json:"url"`
}

type SpotifyUser struct {
	DisplayName string         `json:"display_name"`
	Email       string         `json:"email"`
	Href        string         `json:"href"`
	SpotifyId   string         `json:"id"`
	Images      []SpotifyImage `json:"images"`
}

func GetSpotifyUser(accessToken, refreshToken string) (*SpotifyUser, string, error) {
	body, newAccessToken, err := spotifyRequest(accessToken, refreshToken, "https://api.spotify.com/v1/me")
	if err != nil {
		return nil, "", err
	}

	var spotifyUser SpotifyUser
	if err := json.Unmarshal(body, &spotifyUser); err != nil {
		return nil, "", err
	}

	return &spotifyUser, newAccessToken, nil
}

File: lib/providers/spotify/getSpotifyUser.ts
"use server";
import { createJwt } from "@/lib/jwt";
import { cookies } from "next/headers";
import spotifyRequest from "./spotifyRequest";
import { SpotifyUser } from "@/types/spotifyTypes";
import { AUTH_COOKIE } from "@/cookies/cookieNames";
import cookieStoreToAuthJwtRes from "@/cookies/cookieStoreToAuthJwtRes";

export default async function getSpotifyUser() {
  const cookieStore = await cookies();
  const authRes = cookieStoreToAuthJwtRes(cookieStore);

  if (!authRes.verified) {
    return null;
  }

  const res = await spotifyRequest(
    {
      accessToken: authRes.claims.accessToken,
      refreshToken: authRes.claims.refreshToken,
    },
    "https://api.spotify.com/v1/me",
  );

  if (!res) {
    return null;
  }

  const data = (await res.data.json()) as SpotifyUser;

  const newJwt = createJwt({
    ...authRes.claims,
    picture: data.images[0].url,
    name: data.display_name,
  });
  cookieStore.set(AUTH_COOKIE, newJwt);

  return data;
}
